st2common/st2common/util/secrets.py

```
@@ -27,36 +27,146 @@
+                secret_parameters.append(parameter_type)
+            elif isinstance(secret_parameters, dict):
+                secret_parameters[parameter] = parameter_type
+            else:
+                return parameter_type
 
     return secret_parameters
 
 
-def mask_secret_parameters(parameters, secret_parameters):
+def mask_secret_parameters(parameters, secret_parameters, result=None):
     """
     Introspect the parameters dict and return a new dict with masked secret
     parameters.
-
     :param parameters: Parameters to process.
-    :type parameters: ``dict``
-
-    :param secret_parameters: List of parameter names which are secret.
-    :type secret_parameters: ``list``
+    :type parameters: ``dict`` or ``list`` or ``string``
+
+    :param secret_parameters: Dict of parameter names which are secret.
+                              The type must be the same type as ``parameters``
+                              (or at least behave in the same way),
+                              so that they can be traversed in the same way as
+                              recurse down into the structure.
+    :type secret_parameters: ``dict``
+
+    :param result: Deep copy of parameters so that parameters is not modified
+                   in place. Default = None, meaning this function will make a
+                   deep copy before starting.
+    :type result: ``dict`` or ``list`` or ``string``
     """
-    result = copy.deepcopy(parameters)
-
-    for parameter in secret_parameters:
-        if parameter in result:
-            result[parameter] = MASKED_ATTRIBUTE_VALUE
+    # how we iterate depends on what data type was passed in
+    iterator = None
+    is_dict = isinstance(secret_parameters, dict)
+    is_list = isinstance(secret_parameters, list)
+    if is_dict:
+        iterator = six.iteritems(secret_parameters)
+    elif is_list:
+        iterator = enumerate(secret_parameters)
+    else:
+        return MASKED_ATTRIBUTE_VALUE
```

That's fine or we could also throw because an unsupported type was passed it :)